design/general_refactorings.md

```diff
@@ -0,0 +1,156 @@
+
+##### Recommendation
+
+The change i want to make is to organize the files by their types and not by the command they do and have directories that can be used for the files without directory:
+
+| Directory Name | Purpose                                                      |
+| :------------- | :----------------------------------------------------------- |
+| utils          | wrappers and utils of diffrent python libsssl_servicearg_parse_wrapperhttp_wrapper |
+| api            | files that talk directly with conjur.  conjur_client.py (conjur.py) , conjur_api.py (api.py), ssl_client.py (ssl_service) |
+| logics         | all logic classes here                                       |
+| controllers    | all controller classes                                       |
+| logics         | all logic classes                                            |
+| data_objects   | all data objects classes                                     |
+
+
+
+### Improve Error Handling
+
+------
+
+##### Proposal
+
+Raise only our own exceptions and not built in or libs exceptions. Helps to better understand the error handling flow
+and allows us to catch our own objects and not general exceptions.  In addition all exception should be documented.
+
+##### Current state
+
+Through the codebase, we have very general errors that are not representative of the actual error that took place. The exception is not documented
+For example in `get_certificate` function we do :
+
+```python
+def get_certificate(self, hostname, port):
+    """
+    Method for connecting to Conjur to fetch the certificate chain
+    """
+    if port is None:
+        port = DEFAULT_PORT
+    try:
+        fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
+        logging.debug("Successfully fetched certificate")
+    except Exception as error:
+        raise Exception(f"Unable to retrieve certificate from {hostname}:{port}. " \
+                        f"Reason: {str(error)}") from error
+
+    return fingerprint, readable_certificate
+```
+
+We raise Exception if anything happens in certificate retrivial.
+
+##### Recommendation
+
+Go over error and exception handling and create Conjur-specific exceptions. The code block above should be:
+
+```python
+    def get_certificate(self, hostname, port):
+        """
+        Method for connecting to Conjur to fetch the certificate chain
+        :raises CertificateRetrivialFailed when certificate retrivial failes
+        """
+        if port is None:
+            port = DEFAULT_PORT
+        try:
+            fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
+            logging.debug("Successfully fetched certificate")
+        except Exception as error:
```

Comment:
Yes it would. Can you list which edge cases can happen in 
`fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)`
and which exception can be thrown ?